mailet/standard/src/main/java/org/apache/james/transport/matchers/dlp/DlpDomainRules.java

```diff
@@ -0,0 +1,203 @@
+            return this;
+        }
+
+        private Stream<String> listRecipientsAsString(Mail mail) throws MessagingException {
+            return Stream.concat(listEnvelopRecipients(mail), listHeaderRecipients(mail));
+        }
+
+        private Stream<String> listEnvelopRecipients(Mail mail) {
+            return mail.getRecipients().stream().map(MailAddress::asString);
+        }
+
+        private Stream<String> listHeaderRecipients(Mail mail) throws MessagingException {
+            MimeMessage message = mail.getMessage();
+            if (message != null) {
+                return Arrays.stream(message.getAllRecipients()).map(Address::toString);
+            }
+            return Stream.of();
+        }
+
+        public DlpDomainRulesBuilder senderRule(DLPConfigurationItem.Id id, Pattern pattern) {
+            rules.add(
+                new Rule(
+                    id,
+                    (Mail mail) -> listSenders(mail).anyMatch(pattern.asPredicate())));
+            return this;
+        }
+
+        private Stream<String> listSenders(Mail mail) throws MessagingException {
+            return Stream.concat(listEnvelopSender(mail), listFromHeaders(mail));
+        }
+
+        private Stream<String> listEnvelopSender(Mail mail) {
+            return OptionalUtils.toStream(Optional.ofNullable(mail.getSender()).map(MailAddress::asString));
+        }
+
+        private Stream<String> listFromHeaders(Mail mail) throws MessagingException {
+            MimeMessage message = mail.getMessage();
+            if (message != null) {
+                return Arrays.stream(message.getFrom()).map(Address::toString);
+            }
+            return Stream.of();
+        }
+
+        public DlpDomainRulesBuilder contentRule(DLPConfigurationItem.Id id, Pattern pattern) {
+            rules.add(
+                new Rule(
+                    id,
+                    (Mail mail) -> matchSubject(pattern, mail) || matchBody(pattern, mail)));
+            return this;
+        }
+
+        private boolean matchSubject(Pattern pattern, Mail mail) throws MessagingException {
+            return getMessageSubjects(mail).anyMatch(pattern.asPredicate());
+        }
+
+        private Stream<String> getMessageSubjects(Mail mail) throws MessagingException {
+            MimeMessage message = mail.getMessage();
+            if (message != null) {
+                return Stream.of(message.getSubject());
+            }
+            return Stream.of();
+        }
+
+        private boolean matchBody(Pattern pattern, Mail mail) throws MessagingException, IOException {
+            return getMessageBodies(mail).anyMatch(pattern.asPredicate());
+        }
+
+        private Stream<String> getMessageBodies(Mail mail) throws MessagingException, IOException {
+            MimeMessage message = mail.getMessage();
+            if (message != null) {
+                Object content = message.getContent();
```

Comment:
you don't extract text? You risk to have encoding issues, no?

Reply:
idem for headers unscrambling

Reply:
Yes, I knew that, and benoit handled it.